fix(trim-paths): trim `SO` and `DW_AT_comp_dir` symbols for root DI node

weihanglo opened the pull request:
This is one way to fix <https://github.com/rust-lang/rust/issues/117652>.

## The issue

When `--remap-path-scope=object` is specified, user expect that there is
no local path embedded in final executables. Under `object` scope, the
current implementation only remap debug symbols if debug info is
splitted into its own file. In other words, when
`split-debuginfo=packed|unpacked` is set, rustc assumes there is no
embedded path in the final executable needing to be remapped.

However, this doesn't work.

* On Linux, the root `DW_AT_comp_dir` of a compile unit seems to go into the binary executables.
* On macOS, `SO` symbols are embedded in binary executables and libraries regardless a split-debuginfo file is built. Each `SO` symbol contains a path to the root source file of a debug info compile unit.

## The approach

Path of working directory in the root DI node seems to be embedded in
executables. Hence, we trim them when the scope of `unsplit-debuginfo`
is present, as if it is kinda a "unsplit" debuginfo.


## Unresolved issue

* Not sure where to add more test to consolidate it.
* Haven't investigate if we should apply the same logic to cranelift [here](https://github.com/rust-lang/rust/blob/64d7e0d0b61c460fbc882ae37c0f236756dd9c39/compiler/rustc_codegen_cranelift/src/debuginfo/mod.rs#L68-L80).
* Not sure if there is any other consequence doing this, but AFAIK debugging still works on macOS and Linux  with `profile.dev.trim-paths="object"` fairly (with some extra tweaks in cargo).
* This might be related?
    from <https://github.com/llvm/llvm-project/blob/847d8457d16a7334ba39bdd35c70faa1b295304d/clang/lib/CodeGen/CGDebugInfo.cpp#L623-L631>
    ```
    The DIFile used by the CU is distinct from the main source
    file. Its directory part specifies what becomes the
    DW_AT_comp_dir (the compilation directory), even if the source
    file was specified with an absolute path.
    ```

weihanglo commented:
cc @Urgau if you're interested.

weihanglo commented:
Not the right fix…

weihanglo commented:
r? @pnkfelix 

as you're one member of debugging wg
Thanks!

weihanglo commented:
Find a way that can fix issue on both Linux and Darwin.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7145507870/job/19463512840)
Build commit: 4311f2ea0060c6b2a4a6abaaf3199c9e4eed9f7a (`4311f2ea0060c6b2a4a6abaaf3199c9e4eed9f7a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7145507870/job/19463512840"},"merge_sha":"4311f2ea0060c6b2a4a6abaaf3199c9e4eed9f7a"} -->

weihanglo commented:
This is ready for review again :)

kwfn commented on tests/run-make/split-debuginfo/Makefile:
```diff
@@ -11,23 +13,77 @@ off:
+# - Debuginfo in binary file
+# - `.o` deleted
+# - `.dSYM` present
+# - in binary, paths from `N_SO` (source files) and `N_OSO` (object files) shouldn be remapped
```
```suggestion
# - in binary, paths from `N_SO` (source files) and `N_OSO` (object files) shouldn't be remapped
```

weihanglo commented:
@michaelwoerister 

No hurry. Just want to make sure this wasn't accidentally out of your todo list :)

michaelwoerister commented:
Yes, sorry for the delay. I need to get up-to-date with the current version of the code. At a first glance, I would have expected `.for_codegen()` to have the desired effect. But I'm not up-to-date with remapping scopes yet.

weihanglo commented:
Just resolved the conflict (c string literal).

Let us @bors try :)

weihanglo commented:
@bors try

bors commented:
:hourglass: Trying commit e15a22c6e008f87776ef5ba1f0be3e9299afef67 with merge d2d2599faf05c1fde0679d2fba8ffb44d7a97468...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e15a22c6e008f87776ef5ba1f0be3e9299afef67","merge_sha":"d2d2599faf05c1fde0679d2fba8ffb44d7a97468"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7505429169/job/20437253063)
Build commit: d2d2599faf05c1fde0679d2fba8ffb44d7a97468 (`d2d2599faf05c1fde0679d2fba8ffb44d7a97468`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7505429169/job/20437253063"},"merge_sha":"d2d2599faf05c1fde0679d2fba8ffb44d7a97468"} -->

michaelwoerister commented on compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:
```diff
@@ -852,9 +852,22 @@ pub fn build_compile_unit_di_node<'ll, 'tcx>(
+
+    // Path of working directory in the root DI node seems to be embedded in
+    // executables. Hence, we trim them when the scope of `unsplit-debuginfo`
+    // is present, as if it is kinda a "unsplit" debuginfo.
```
I've been able to verify that this is indeed the case for DWARF dwps (i.e. the DW_AT_comp_dir is only contained in the binary, not in the dwp). However, I'm not sure that this behavior is guaranteed, so I think we should test that the `dwp` and `dwo` files don't contain the compiler's working directory at all.

michaelwoerister commented on tests/run-make/split-debuginfo/Makefile:
```diff
@@ -11,23 +13,77 @@ off:
+	dsymutil -s $(TMPDIR)/foo | grep $(TMPDIR) || exit 1 # expected: `grep $(TMPDIR)` to exit 1
+	dsymutil -s $(TMPDIR)/foo | (! grep $(HERE)) || exit 1
+	rm -rf $(TMPDIR)/foo.dSYM
+	rm $(TMPDIR)/$(call BIN,foo)
```
I'd suggest changing all of the remapping tests in this file to do a sanity check:
- remap paths to something that is unlikely to be contained in the binary by accident (e.g. `/my-remapped-src` and `/my-remapped-cwd` instead of `/a` and `/b`)
- check that `/my-remapped-src` and `/my-remapped-cwd` *are* contained in the various files, so that we make sure the tests would actually fail if something is wrong. These run-make tests are very prone to bugs where the tests silently pass without testing anything.

weihanglo replied:
Updated, but I don't think it is in a good state to merge.


If you look closer at the changes in tests in the last commit, you can see it is like a whack-a-mole game.

* Some symbols in split debuginfo file got overly remapped in `object` scope, and some should be remapped but didn't.
* So do unpacked debuginfo in `.o` files.
* `.debug-line` doesn't seem to be handled well.
* When unpacked `.o` object files contain debuginfo, do we consider them split debuginfo files?


That leads to a question: Does llvm provide such a granular API for rustc to implement RFC 3127? 

(or maybe this is not a task for me 😞)

michaelwoerister replied:
I haven't taken an in-depth look but [DIBuilder::createFile](https://llvm.org/doxygen/classllvm_1_1DIBuilder.html#afac26b4f6298753fa1566e66758c9dbe) does not seem to offer anything.

I think it's time to re-evaluate this particular aspect of RFC 3127. Maybe we just need to remove the split-/unsplit-debuginfo distinction and just have a single `debuginfo` scope. That still seems practical to me. I've created a todo for me to post something to the tracking issue. Let's mark the PR as blocked for now.

weihanglo replied:
Thanks for the review, Michael!

What I am worried about is that scope `object` and `debug` never actually work as defined in RFC 3127 unless we have some post-processing approaches. Reading through [this post](https://reproducible-builds.org/docs/build-path/) and other tools like clang and gcc, seems none of them offers such an option to deal with split-debuginfo files and binaries separately like RFC 3127 suggesting.

michaelwoerister commented on tests/run-make/split-debuginfo/Makefile:
```diff
@@ -141,9 +205,12 @@ packed-remapped-single:
 # - `.dwp` present
 packed-remapped-scope:
```
We should add a test with scope `split-debuginfo,split-debuginfo-path` that greps the entire `.dwp` file for remapped paths (i.e. without going through objdump or readelf).

michaelwoerister reviewed (commented):
Thanks again for the PR, @weihanglo! 

It looks to me, like it does the correct thing. I've added a few suggestions on how to make the tests more robust.

I'm not quite clear yet on the situation on macOS. Do you have a link to some documentation on the various SO symbols?

michaelwoerister commented:
By the way, it looks to my like the [RemapFileNameExt::for_scope](https://github.com/rust-lang/rust/blob/73252d51d13c0e4f872c3c51defcf48174860ef6/compiler/rustc_session/src/session.rs#L1526) implementations should all use `.intersects()` instead of `.contains()`, right?

Urgau commented:
> By the way, it looks to my like the [RemapFileNameExt::for_scope](https://github.com/rust-lang/rust/blob/73252d51d13c0e4f872c3c51defcf48174860ef6/compiler/rustc_session/src/session.rs#L1526) implementations should all use `.intersects()` instead of `.contains()`, right?

I don't think so, the goal the `for_scope` function is to return a local or remapped filename if the scope passed was enabled not if _some part_ of the scope was enabled [\[1\]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=44d8f1ed5bbcdd7a185722c4205a4a70).

Otherwise only enabling the `MACRO` scope and checking for the presence of the `OBJECT` scope (`MACRO | UNSPLIT_DEBUGINFO | SPLIT_DEBUGINFO_PATH` [\[2\]](https://github.com/rust-lang/rust/blob/73252d51d13c0e4f872c3c51defcf48174860ef6/compiler/rustc_session/src/config.rs#L1007-L1032)) would return a remapped filename, which is not what we want.

michaelwoerister commented:
> don't think so, the goal the for_scope function is to return a local or remapped filename if the scope passed was enabled not if some part of the scope was enabled.

Here is my reasoning for why I think this should be `.intersect()`:

1. We have the following method:  
   ```rust
   fn for_scope(&self, sess: &Session, scopes: RemapPathScopeComponents) -> Self::Output<'_> {
         if sess.opts.unstable_opts.remap_path_scope.contains(scopes) {
             self.prefer_remapped_unconditionaly()
         } else {
             self.prefer_local()
         }
   }
   ```
  
   where `sess.opts.unstable_opts.remap_path_scope` denotes which scopes should be remapped and `scopes` denotes all the scopes a given filepath belongs too. For example, if the path will be emitted to split-debuginfo, then `scopes` will be `SPLIT_DEBUGINFO`, if the path will end up in both in split-debuginfo and unsplit-debuginfo (for example because LLVM will just emit that single string into both artifacts), then `scopes` will be `SPLIT_DEBUGINFO | UNSPLIT_DEBUGINFO`.
2. If the given path will only end up in a single scope (which is usually the case), then `.contains()` and `.intersects()` will yield the same result, so it does not matter.
3. If the path ends up in multiple scopes, some of which are enabled and some of which aren't, then we have to decide how to deal with that. We can either only apply the remapping if all scopes are enabled (= `.contains()`) or apply the remapping if any of the scopes are enabled (= `.intersects()`). In both cases we don't have a perfect match. However, I would argue that we should definitely err on the side of applying the remapping too much rather than too little, because remapping is used for privacy reasons. Otherwise we could end up with private information being embedded even though a seemingly correct remapping is specified.

Alternatively, we could assert that `scopes` only has a single bit set. I'm not sure if there are cases where a given path ends up in two different places (I thought that would be the case for the `DW_AT_comp_dir` but that turned out to be wrong).

Urgau commented:
I'm not sure I follow your logic for the `for_scope` function. In particular this part:

> `scopes` denotes all the scopes a given filepath belongs too.

Does this mean that you associate a filepath with `scopes`? Because that not how I see things. 

When I implemented the RFC in https://github.com/rust-lang/rust/pull/115214 I just made all the places that remapped paths to be conditional on the scope(s) they applied to (when emitting a diagnostic it's `DIAGNOSTICS`, for macro it's `MACROS`, ...). A filepath is not associated with any scopes, it's remapped based on the scope of the area that wants to "emit it".

> For example, if the path will be emitted to split-debuginfo, then `scopes` will be `SPLIT_DEBUGINFO`

The variable `scopes` is not user-defined, it's defined in the code for each specific location. Only `sess.opts.unstable_opts.remap_path_scope` is user-defined.

> if the path will end up in both in split-debuginfo and unsplit-debuginfo (for example because LLVM will just emit that single string into both artifacts), then `scopes` will be `SPLIT_DEBUGINFO | UNSPLIT_DEBUGINFO`.

`scopes` cannot currently ever be that combination. (well technically yes, as nothing prevents another dev to test for it but it's not the case currently)

> Alternatively, we could assert that `scopes` only has a single bit set.

Well that's currently the case, it's only ever called with the [`MACRO` and `DIAGNOSTICS` scopes](https://github.com/search?q=repo%3Arust-lang%2Frust+%2Ffor_scope%5C%28.*%2C+RemapPathScopeComponents%3A%3A%2F&type=code). So we could assert this in the function.

michaelwoerister commented:
> Does this mean that you associate a filepath with scopes? Because that not how I see things.

>   
> When I implemented the RFC in https://github.com/rust-lang/rust/pull/115214 I just made all the places that remapped paths to be conditional on the scope(s) they applied to (when emitting a diagnostic it's DIAGNOSTICS, for macro it's MACROS, ...). A filepath is not associated with any scopes, it's remapped based on the scope of the area that wants to "emit it".

I think I might have formulated things in a confusing way. With "filepath" I just mean the path that is being remapped, that is, the thing that implements `RemapFileNameExt`.

> A filepath is not associated with any scopes, it's remapped based on the scope of the area that wants to "emit it".

I'm not sure I agree with that (although the distinction is subtle). In my mind, we use the concept of "scopes" to model where a given path will be emitted to. For example, if we are about to emit some path as part of a diagnostic message, the we use `.for_scope(DIAGNOSTICS)`. It's not `DIAGNOSTICS` because it is being emitted as part of some diagnostics machinery -- it is `DIAGNOSTICS` because it will show up in diagnostic output.

By extension, if we are storing a path to some place (like debuginfo in LLVM IR) that might then end up in multiple output artifacts and these output artifacts are in different scopes (e.g. LLVM might write it both to the object file and to the `.dwo`), then the corresponding `.for_scope()` needs to take the union of all scopes that the path will end up in. This is the reason why something like `SPLIT_DEBUGINFO | UNSPLIT_DEBUGINFO` could validly occur.

weihanglo commented on tests/run-make/split-debuginfo/Makefile:
```diff
@@ -11,23 +13,77 @@ off:
+# - Debuginfo in binary file
+# - `.o` deleted
+# - `.dSYM` present
+# - in binary, paths from `N_SO` (source files) and `N_OSO` (object files) shouldn be remapped
```
Hey, @michaelwoerister. I've move <https://github.com/rust-lang/rust/pull/118518#pullrequestreview-1821356351> to this thread so we can follow the discussion better.

> I'm not quite clear yet on the situation on macOS. Do you have a link to some documentation on the various SO symbols?

It's also unclear to me. I have no idea if there is an official doc for Mach-O debug symbols. Here is what I've found:

* [Apple's “Lazy” DWARF Scheme](https://wiki.dwarfstd.org/Apple%27s_%22Lazy%22_DWARF_Scheme.md)
* [Apple's Linker & Deterministic Builds](https://milen.me/writings/apple-linker-ld64-deterministic-builds-oso-prefix/)
* [SymbolTable.pm](https://opensource.apple.com/source/cctools/cctools-973.0.1/tests/include/MachO/SymbolTable.pm) from old Apple open source header (search for `N_SO` and `N_OSO`)
* [ArchiveWriter.cpp](https://github.com/llvm/llvm-project/blob/e81c981fe35483a4b7474e110ff1b5d02bc2cb00/llvm/lib/Object/ArchiveWriter.cpp#L738-L751) from LLVM
* [My own experiments on it](https://github.com/rust-lang/rust/issues/116948#issuecomment-1793617018)

michaelwoerister replied:
I'll try to read up on those links by the end of the week. If we can't get to a clear conclusion, I think we should merge the PR nonetheless, for the improved behavior it provides on Linux.

michaelwoerister replied:
OK, from reading through those links it sounds to me like:

- SO symbols we can handle fine by remapping the paths generated by debuginfo.
- OSO symbols are added by the linker. They are not needed after dsymutil has built the dSYM bundle. So it would be fine to remove them after building the dSYM bundle.

I think the best option would be to use the `-oso_prefix` linker option. My guess is that will not be a problem even on older macOS versions. But I don't know if we have a minimum supported XCode version (in addition to a minimum support macOS version).

Other than that, rewriting/removing the OSO symbols in a postprocessing step seems to be our only option.

michaelwoerister replied:
@weihanglo, do you know how `-Cstrip=debuginfo` behaves on macOS? Does that remove the OSO symbols?

weihanglo replied:
It runs `strip -S` on macOS and removes OSO symbols successfully, and the `.dSYM` bundle can successfully find sources for a simple `hello-world.rs` example.

Should we always pass `-Cstrip=debuginfo` to deal with OSO symbols? I am not sure…

michaelwoerister replied:
It looks like `-Cstrip=debuginfo` is [now the default](https://github.com/rust-lang/cargo/pull/13257) unless debuginfo is requested. We could also do a similar thing for when trim-paths is enabled.

weihanglo replied:
That sounds good, but only when Cargo is involved. OSO is still there if using rustc directly.

(granted, people invoking rustc directly should be able to set `-Cstript=debuginfo`)

weihanglo replied:
or do you mean that rustc should implicitly strip debuginfo under some trim-paths circumstances?

michaelwoerister replied:
I was more thinking about the Cargo case. For direct rustc invocations, the defaults are less important, I think, because that's more of an advanced use case. The test cases here could simulate what we expect Cargo to do.

michaelwoerister commented:
> Well that's currently the case, it's only ever called with the [MACRO and DIAGNOSTICS scopes](https://github.com/search?q=repo%3Arust-lang%2Frust+%2Ffor_scope%5C%28.*%2C+RemapPathScopeComponents%3A%3A%2F&type=code). So we could assert this in the function.

@Urgau, if you are interested in opening a PR that asserts that only a single scope is passed to for_scope, and that also adds a docstring saying that only a single scope is supported, you can r? me. Since for_scope currently accepts a set of scopes, I think it is a bit confusing at the moment. But a bit of documentation and an assertion, would be an easy fix for that.

weihanglo committed:
test: verify current remap rules on darwin

When `--remap-path-scope=object` is specified, user expect that there is
no local path embedded in final executables. Under `object` scope, the
current implementation only remap debug symbols if debug info is
splitted into its own file. In other words, when
`split-debuginfo=packed|unpacked` is set, rustc assumes there is no
embedded path in the final executable needing to be remapped.

However, this doesn't work on macOS. On macOS, `SO` symbols are embedded
in binary executables and libraries regardless a split-debuginfo file is
built. Each `SO` symbol contains a path to the root source file of a
debug info compile unit.

This commit demonstrates the case, and hope there is a fix soon.

weihanglo committed:
test: verify current remap rules on Linux

When `--remap-path-scope=object` is specified, user expect that there is
no local path embedded in final executables. Under `object` scope, the
current implementation only remap debug symbols if debug info is
splitted into its own file. In other words, when
`split-debuginfo=packed|unpacked` is set, rustc assumes there is no
embedded path in the final executable needing to be remapped.

However, this doesn't work on Linux. On Linux, the root `DW_AT_comp_dir`
of a compile unit seems to go into the binary binary executables.

This commit demonstrates the case, and hope there is a fix soon.

weihanglo committed:
fix(trim-paths): trim working diretory from root DI node

Path of working directory in the root DI node seems to be embedded in
executables. Hence, we trim them when the scope of `unsplit-debuginfo`
is present, as if it is kinda a "unsplit" debuginfo.

michaelwoerister commented:
Marking the PR as blocked until we've discussed whether the RFC can be implemented at all in its current form. See https://github.com/rust-lang/rust/pull/118518#discussion_r1464656443

bors commented:
:umbrella: The latest upstream changes (presumably #122450) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

weihanglo commented:
I am a bit lost. @michaelwoerister, this is still worth pursuing, or at least I should verify the current situation again, right?

Urgau commented:
Some of the test changes look worth landing, even if the logic change is no longer necessary.

Better have more tests than not enough.

michaelwoerister commented:
It's definitely worth having tests for the trim-paths related functionality. Maybe its better to create a number of smaller test cases. It might also be possible to use regular UI tests for this. I recently saw something pretty clever in another PR where a UI test looks for its own PDB file:
https://github.com/rust-lang/rust/blob/95adb6d7b520230971522c8806cc6ab518e419c5/tests/ui/debuginfo/msvc-strip-debuginfo.rs

We could do the same, i.e. just load binaries and separate debuginfo files into memory and then treat them as byte slices where we look for the text we want to find (or don't want to find). That should be a lot more readable than run-make tests. If compiletest allows it, we can also put some shared code into a common aux-crate.

Regarding what to test: I suggest picking the most relevant configurations from [the table in the tracking issue](https://github.com/rust-lang/rust/issues/111540#issuecomment-1994010274) and then have a test for linux, macOS, and MSVC. For each useful `(platform, trim-paths, split-debuginfo)` combination, we compile with the commandline options that we expect Cargo to pass for that combination and then make sure the binaries and separate debuginfo files are either sanitized or not sanitized. I don't think we need to test the less useful combinations. 

It might be good to split this into a number of smaller PRs, starting with a simple case just for Linux or macOS. Feel free to assign to me for review.

weihanglo commented:
Currently don't have time working on this. It seems that we need a complete rewrite because runmake is under the gigantic porting. So close.

michaelwoerister commented:
Thanks, @weihanglo! This PR is triggered a lot of valuable work around clarifying the RFC. We'll get there 🙂
